misc/testerina/modules/testerina-core/src/main/ballerina/execute.bal

```diff
@@ -36,56 +40,107 @@ public function startSuite() returns int {
     } else {
         if testRegistry.getFunctions().length() == 0 && testRegistry.getDependentFunctions().length() == 0 {
             println("\tNo tests found");
-            return exitCode;
+            lock {
+                return exitCode;
+            }
+
         }
 
         error? err = orderTests();
         if err is error {
-            exitCode = 1;
+            enableExit();
             println(err.message());
         } else {
             executeBeforeSuiteFunctions();
-            executeTests();
+            err = executeTests();
+            if err is error {
+                enableExit();
+                println(err.message());
+            }
             executeAfterSuiteFunctions();
             reportGenerators.forEach(reportGen => reportGen(reportData));
         }
     }
-    return exitCode;
+    lock {
+        return exitCode;
+    }
 }
 
-function executeTests() {
+function executeTests() returns error? {
+    decimal startTime = currentTimeInMillis();
     foreach TestFunction testFunction in testRegistry.getFunctions() {
-        executeTest(testFunction);
+        _ = testFunction.parallelizable ? conMgr.addInitialParallelTest(testFunction) : conMgr.addInitialSerialTest(testFunction);
+    }
+
+    while !conMgr.isExecutionDone() {
+
+        if conMgr.getAvailableWorkers() != 0 {
+            conMgr.waitUntilEmptyQueueFilled();
+
+            if conMgr.getSerialQueueLength() != 0 && conMgr.getAvailableWorkers() == testWorkers {
+                TestFunction testFunction = conMgr.getSerialTest();
+                conMgr.addTestInExecution(testFunction);
+                conMgr.allocateWorker();
+                future<error?> serialWaiter = start executeTest(testFunction);
```

Comment:
If `executeTests()` is not an `isolated` function, these will all get scheduled on strands of the same thread. Is that intentional?

Reply:
We are currently using same function(i.e `executeTest(testFunction)`) to execute both isolated, non-isolated test functions. If we make it isolated, can we use same function(i.e `executeTest(testFunction)`) to execute both isolated, non-isolated test functions?

Reply:
You can't call a non-isolated function is an isolated function. So you'll have to separate the logic out.

Reply:
So, We may need to have 2 separate versions(isolated and non-isolated) of same logic. Do we have any other ways for this kind of scenarios??